### PR TITLE
Add socks_proxy support for ftp/sftp/ssh backend

### PR DESCRIPTION
**Pull request by snowzach**
#### What is the purpose of this change?
Add support for socks 5 proxies to the SFTP/SSH backend

#### Was the change discussed in an issue or in the forum before?

https://github.com/rclone/rclone/issues/3558

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [X] I have added tests for all changes in this PR if appropriate.
- [X] I have added documentation for the changes if appropriate.
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [X] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: backend/ftp/ftp.go [hunk 1]
```diff
@@ -174,6 +175,18 @@ Enabled by default. Use 0 to disable.`,
 If this is set and no password is supplied then rclone will ask for a password
 `,
 			Advanced: true,
+		}, {
+			Name:    "socks_proxy",
+			Default: "",
+			Help: `Socks 5 proxy host for both SSH and SFTP connections.
```
Comment (ncw) [inline backend/ftp/ftp.go:181]: This should just say `Socks 5 proxy host.` - it is for the FTP backend so the ssh part isn't relevant.

File: backend/sftp/sftp.go [hunk 2]
```diff
@@ -388,6 +390,18 @@ Example:
     ssh-ed25519 ssh-rsa ssh-dss
 `,
 			Advanced: true,
+		}, {
+			Name:    "socks_proxy",
+			Default: "",
+			Help: `Socks 5 proxy host for both SSH and SFTP connections.
```
Comment (ncw) [inline backend/sftp/sftp.go:396]: Again just `Socks 5 proxy host.`

File: backend/sftp/sftp.go [hunk 3]
```diff
@@ -467,8 +482,36 @@ type Object struct {
...
+			}
+			if len(proxyCreds) == 2 {
+				proxyAuth.Password = proxyCreds[1]
+			}
+			proxyAddress = credsAndHost[1]
+		} else {
+			proxyAddress = credsAndHost[0]
+		}
+		// Build the proxy dialer
+		proxyDialer, _ := proxy.SOCKS5("tcp", proxyAddress, proxyAuth, baseDialer)
```
Comment (ncw) [inline backend/sftp/sftp.go:508]: Don't ignore the error please!

File: backend/ftp/ftp.go [hunk 4]
```diff
@@ -359,7 +373,31 @@ func (f *Fs) ftpConnection(ctx context.Context) (c *ftp.ServerConn, err error) {
...
+				}
+				if len(proxyCreds) == 2 {
+					proxyAuth.Password = proxyCreds[1]
+				}
+				proxyAddress = credsAndHost[1]
+			} else {
+				proxyAddress = credsAndHost[0]
+			}
+			// Build the proxy dialer
+			proxyDialer, _ := proxy.SOCKS5("tcp", proxyAddress, proxyAuth, baseDialer)
```
Comment (ncw) [inline backend/ftp/ftp.go:395]: Don't ignore error.

**Review by ncw — COMMENTED** [review]
Thanks for doing this - this has been an often requested feature.

See comments inline.

Perhaps rclone should be picking up the SOCKS5 proxy from environment vars like we do for the http services?

https://rclone.org/faq/#can-i-use-rclone-with-an-http-proxy

Are there standard env vars we could/should use instead of the backend configs?

Also the code duplicated between the sftp and ftp backends is slightly annoying! Maybe it could be factored into lib/proxy or something like that? I suspect the code might grow and its just on the edge of too much duplication for my tastes!

Thank you

File: backend/ftp/ftp.go (hunk 1 above)
Comment (snowzach) [reply backend/ftp/ftp.go]: Updated

File: backend/sftp/sftp.go (hunk 2 above)
Comment (snowzach) [reply backend/sftp/sftp.go]: Updated

File: backend/sftp/sftp.go (hunk 3 above)
Comment (snowzach) [reply backend/sftp/sftp.go]: I am not sure why the function returns an error. It doesn't actually trigger one.. but it looks bad so I handled it..

File: backend/ftp/ftp.go (hunk 4 above)
Comment (snowzach) [reply backend/ftp/ftp.go]: Same as above

**Comment by snowzach** [comment]
Yeah, makes sense to make proxy dialing it's own package. I moved it to `lib/proxy` in case you want to do more with it someday.

**Review by ncw — APPROVED** [review]
Looks very nice now 😀

What did you think of the idea of picking up environment variables?

I can merge this and we can discuss further or you could update this , whichever you like.

**Comment by snowzach** [comment]
I think this one probably stands pretty good on it's own. I can see cases where you want to configure a socks proxy on it's own for one connection and not everything (which is actually what I need for my use case). The other option being implementing something globally on your `fshttp.NewDialer` function. That's gonna effect everything that uses it though. 

I would be willing to work on that if you think it's needed. The only danger being if you configure both... I am sure that could be guarded against though. 

I'd be happy to look into that for you if you've received enough requests for it. I'll back out these changes if it ends up superseding them. (Perhaps an option for NewDialer with the socks proxy endpoint. If not populated defaults to global config which may or may not be for a proxy) I actually started that way but it was a bit messy because there's a lot of Dial/DialContext overlap/usage. For whatever reason the proxy dialer doesn't implement DialContext so you have to create a wrapper for it.

**Comment by ncw** [comment]
OK I will merge this now and we can wait and see if anyone requests environment variables. I'm not sure there is a standard for them - here is the best doc I found

https://gist.github.com/yougg/5d2b3353fc5e197a0917aae0b3287d64#set-ssh-proxy-environment-variables

> The other option being implementing something globally on your `fshttp.NewDialer` function. That's gonna effect everything that uses it though.

I think we already support that for SOCKS5 proxies. I may be wrong though! That is HTTP only backends though.

Thank you for the contribution :-)

**Comment by SaladinoGW** [comment]
Thanks guys, I'm already using this new feature and I imagine many people are too

**Comment by BunnySakura** [comment]
以防有人不知道，使用`rclone config`配置sftp后端，在高级设置找到`socks_proxy`，配置为代理地址即可。配置完成会在配置文件增加一行

```ini
socks_proxy = <IP地址>:<端口>
````

**注意**：低版本rclone不支持此设置

---

Just in case anyone doesn't know, use `rclone config` to configure the sftp backend, find `socks_proxy` in the advanced settings and configure it as a proxy address. When you're done, you'll add a line to the config file

```ini
socks_proxy = <IP>:<port>
````

**Note**: Lower versions of rclone do not support this setting.

**Pull request merged**